src/library_manager/lib_manager.c

```diff
@@ -110,79 +137,135 @@ static int lib_manager_load_module(uint32_t module_id, struct sof_man_module *mo
+				   struct sof_man_fw_desc *desc, uint32_t lib_id)
 {
 	struct ext_library *ext_lib = ext_lib_get();
-	uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
-	void __sparse_cache *va_base_text = (void __sparse_cache *)
-		mod->segment[SOF_MAN_SEGMENT_TEXT].v_base_addr;
-	size_t st_text_size = mod->segment[SOF_MAN_SEGMENT_TEXT].flags.r.length;
-	void __sparse_cache *va_base_rodata = (void __sparse_cache *)
-		mod->segment[SOF_MAN_SEGMENT_RODATA].v_base_addr;
-	size_t st_rodata_size = mod->segment[SOF_MAN_SEGMENT_RODATA].flags.r.length;
-	int ret;
-
-	st_text_size = st_text_size * PAGE_SZ;
-	st_rodata_size = st_rodata_size * PAGE_SZ;
-
-	ret = sys_mm_drv_unmap_region((__sparse_force void *)va_base_text, st_text_size);
-	if (ret < 0)
-		return ret;
-
-	ret = sys_mm_drv_unmap_region((__sparse_force void *)va_base_rodata, st_rodata_size);
-	if (ret < 0)
-		return ret;
 
 	/* There are modules marked as lib_code. This is code shared between several modules inside
 	 * the library. Unload all lib_code modules with last none lib_code module unload.
 	 */
 	if (mod->type.lib_code)
-		return ret;
+		return;
 
 	if (!mod->type.lib_code && ext_lib->mods_exec_load_cnt > 0)
 		ext_lib->mods_exec_load_cnt--;
 
 	if (ext_lib->mods_exec_load_cnt == 0) {
 		struct sof_man_module *module_entry =
 		    (struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(0));
+
 		for (size_t idx = 0; idx < desc->header.num_module_entries;
 				++idx, ++module_entry) {
 			if (module_entry->type.lib_code) {
-				ret =
 				lib_manager_unload_module(lib_id << LIB_MANAGER_LIB_ID_SHIFT |
 							  idx, mod, desc);
 			}
 		}
 	}
+}
+
+static int lib_manager_load_module(uint32_t module_id, struct sof_man_module *mod,
+				   struct sof_man_fw_desc *desc)
+{
+	uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
+	struct lib_manager_mod_ctx *ctx = lib_manager_get_mod_ctx(module_id);
+	uint8_t *load_base = (uint8_t *)ctx->desc;
+	void __sparse_cache *va_base_text = (void __sparse_cache *)
+		mod->segment[SOF_MAN_SEGMENT_TEXT].v_base_addr;
+	void *src_txt = (void *)(load_base + mod->segment[SOF_MAN_SEGMENT_TEXT].file_offset);
+	size_t st_text_size = ctx->segment_size[SOF_MAN_SEGMENT_TEXT];
+	void __sparse_cache *va_base_rodata = (void __sparse_cache *)
+		mod->segment[SOF_MAN_SEGMENT_RODATA].v_base_addr;
+	void *src_rodata = (void *)(load_base +
+				    mod->segment[SOF_MAN_SEGMENT_RODATA].file_offset);
+	size_t st_rodata_size = ctx->segment_size[SOF_MAN_SEGMENT_RODATA];
+	int ret;
+
+	/* Copy Code */
+	ret = lib_manager_load_data_from_storage(va_base_text, src_txt, st_text_size,
+						 SYS_MM_MEM_PERM_RW | SYS_MM_MEM_PERM_EXEC);
+	if (ret < 0)
+		return ret;
+
+	/* .text contains instructions and it also often contains local data */
+	dcache_writeback_region(va_base_text, st_text_size);
+	icache_invalidate_region(va_base_text, st_text_size);
+
+	/* Copy RODATA */
+	ret = lib_manager_load_data_from_storage(va_base_rodata, src_rodata,
```

Comment:
Lack of support for readonly segment and empty data segment.

Reply:
@softwarecki sorry, is this something that this commit changes? That's how it is now, isn't it?